python/tskit/trees.py

```diff
@@ -304,213 +238,207 @@ def span(self):
-        :attr:`TreeSequence.num_mutations` - 1.
-    :vartype id: int
-    :ivar site: The integer ID of the site that this mutation occurs at. To obtain
-        further information about a site with a given ID use
-        :meth:`TreeSequence.site`.
-    :vartype site: int
-    :ivar node: The integer ID of the first node that inherits this mutation.
-        To obtain further information about a node with a given ID, use
-        :meth:`TreeSequence.node`.
-    :vartype node: int
-    :ivar time: The occurrence time of this mutation.
-    :vartype time: float
-    :ivar derived_state: The derived state for this mutation. This is the state
-        inherited by nodes in the subtree rooted at this mutation's node, unless
-        another mutation occurs.
-    :vartype derived_state: str
-    :ivar parent: The integer ID of this mutation's parent mutation. When multiple
-        mutations occur at a site along a path in the tree, mutations must
-        record the mutation that is immediately above them. If the mutation does
-        not have a parent, this is equal to the :data:`NULL` (-1).
-        To obtain further information about a mutation with a given ID, use
-        :meth:`TreeSequence.mutation`.
-    :vartype parent: int
-    :ivar metadata: The decoded :ref:`metadata <sec_metadata_definition>` for this
-        mutation.
-    :vartype metadata: object
+    __slots__ = ["id", "site", "node", "derived_state", "parent", "metadata", "time"]
+    id: int  # noqa A003
+    """
+    The integer ID of this mutation. Varies from 0 to
+    :attr:`TreeSequence.num_mutations` - 1.
+    """
+    site: int
+    """
+    The integer ID of the site that this mutation occurs at. To obtain
+    further information about a site with a given ID use :meth:`TreeSequence.site`.
+    """
+    node: int
+    """
+    The integer ID of the first node that inherits this mutation.
+    To obtain further information about a node with a given ID, use
+    :meth:`TreeSequence.node`.
+    """
+    derived_state: str
+    """
+    The derived state for this mutation. This is the state
+    inherited by nodes in the subtree rooted at this mutation's node, unless
+    another mutation occurs.
+    """
+    parent: int
+    """
+    The integer ID of this mutation's parent mutation. When multiple
+    mutations occur at a site along a path in the tree, mutations must
+    record the mutation that is immediately above them. If the mutation does
+    not have a parent, this is equal to the :data:`NULL` (-1).
+    To obtain further information about a mutation with a given ID, use
+    :meth:`TreeSequence.mutation`.
+    """
+    metadata: Any
+    """
+    The decoded :ref:`metadata <sec_metadata_definition>` for this
+    mutation.
+    """
+    time: float
+    """
+    The occurrence time of this mutation.
     """
 
+    # To get default values on a frozen instance you have work around the freeze
     def __init__(
```

Comment:
Would the post_init hook be better?

Reply:
Possibly - would still need to use the ugly setattr.